python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -87,26 +86,89 @@
 ASSET_SUBSET_INPUT_PREFIX = "__subset_input__"
 
 
-@dagster_model
```

Comment:
Sad to make this no longer a dagster_model. The reason I'm changing it is to compute a derived attribute (`_output_names_by_key`) during construction, while validating that that derived attribute makes sense.

A couple potential alternatives if we want to keep it a dagster_model:
- Compute it in `__new__` and set it after construction
```python
def __new__(...):
    result = super().__new__(...)
    result._output_names_by_key = ...
    return result
```
- Force loading it into the cache in `__new__`
```python
def __new__(...):
    result = super().__new__(...)
    result.output_names_by_key
    return result

@cached_property
def output_names_by_key(self):
    ...
```